### DISCUSSION TITLE
Convert Disposable to Kotlin Flow?

### DISCUSSION THREAD

**Discussion by SeanOThomas** [GetStream/stream-chat-android#1949] [Q&A] [answered]
We want to get real time unread count updates. According to the [docs](https://getstream.io/chat/docs/android/unread/?language=kotlin), we should use the lower-level client `subscribeFor` method which returns a GetStream `Disposable`. It's not ideal for us to expose `Disposable` beyond the data layer.  Is there a convenient or noted way of converting the `Disposable` to Kotlin Flow data? Thanks!

**Comment by zsmb13** [comment, marked as answer]
We don't have API for this yet, but we'll look into providing it.

In the meantime, you should be able to use the [`callbackFlow` function](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/callback-flow.html) from kotlinx.coroutines to convert any callback based API into a `Flow`. For our event API, that would look something like this (haven't tested it, just a basic draft):

```kotlin
val flow = callbackFlow<ChatEvent> {
    val disposable = client.subscribeFor(ChatEventListener { event ->
        this.trySendBlocking(event)
    })
    awaitClose {
        disposable.dispose()
    }
}

// In a coroutine
flow.collect { event: ChatEvent -> 
    // ...
}
```
